ql/src/java/org/apache/hadoop/hive/ql/exec/tez/TezSessionPoolManagerMetrics.java

```diff
@@ -0,0 +1,259 @@
+import org.apache.hadoop.hive.common.metrics.common.MetricsFactory;
+import org.apache.hadoop.hive.common.metrics.common.MetricsVariable;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.util.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+
+/**
+ * TezSessionPoolManagerMetrics collects, aggregates and exposes data from Tez sessions to Hive's own MetricsFactory.
+ * This is independent of the Tez DagClient which is active only while running DAGs.
+ */
+public class TezSessionPoolManagerMetrics {
+  private static final Logger LOG = LoggerFactory.getLogger(TezSessionPoolManagerMetrics.class);
+
+  // metric names to be used by Tez sessions when returning data
+  public static final String TEZ_SESSION_METRIC_RUNNING_TASKS = "tez_session_running_tasks";
+  public static final String TEZ_SESSION_METRIC_PENDING_TASKS = "tez_session_pending_tasks";
+
+  // derived metrics (not set by tez sessions, calculated here)
+  public static final String TEZ_SESSION_METRIC_TASK_BACKLOG_RATIO = "tez_session_task_backlog_ratio";
+  public static final String TEZ_SESSION_METRIC_TASK_BACKLOG_RATIO_PRESENT_SINCE =
+      "tez_session_task_backlog_ratio_%d_present_since";
+  // For scaling decisions, TezSessionPoolManagerMetrics tracks the duration
+  // since the task backlog ratio exceeded a certain threshold.
+  private static final Integer[] TRACKED_TASK_BACKLOG_RATIOS = {1, 2, 5, 10, 20, 50, 100};
+
+  private final TezSessionPoolManager poolManager;
+  private final Metrics metrics;
+  private ScheduledExecutorService mainExecutor;
+
+  // session pool metrics
+  @VisibleForTesting
+  final TezSessionMetric<Integer> runningTasksCount = new TezSessionMetric<>(TEZ_SESSION_METRIC_RUNNING_TASKS, 0);
+  @VisibleForTesting
+  final TezSessionMetric<Integer> pendingTasksCount = new TezSessionMetric<>(TEZ_SESSION_METRIC_PENDING_TASKS, 0);
+  @VisibleForTesting
+  final TezSessionMetric<Double> taskBacklogRatio = new TezSessionMetric<>(TEZ_SESSION_METRIC_TASK_BACKLOG_RATIO, 0.0);
+  private final TaskBacklogRatioMetric[] taskBacklogRatioDurations = Arrays.stream(TRACKED_TASK_BACKLOG_RATIOS)
+      .map(
+          ratio -> new TaskBacklogRatioMetric(String.format(TEZ_SESSION_METRIC_TASK_BACKLOG_RATIO_PRESENT_SINCE, ratio),
+              ratio))
+      .toArray(TaskBacklogRatioMetric[]::new);
+
+
+  public TezSessionPoolManagerMetrics(TezSessionPoolManager poolManager) {
+    this.poolManager = poolManager;
+    this.metrics = MetricsFactory.getInstance();
```

Comment:
should we `init` first?

Reply:
I don't think so, if it's not initialized by default, then it means HS2 is not configured to utilize metrics, in this case, this whole thing isn't expected to be used
this is what I handle by:
```
  public TezSessionPoolManagerMetrics start(HiveConf conf) {
    if (metrics == null) {
      LOG.warn(
          "Metrics are not enabled, cannot start TezSessionPoolManagerMetrics (check {})",
          HiveConf.ConfVars.HIVE_SERVER2_METRICS_ENABLED.varname);
      return this;
    }
```